File: templates/python-agent/stream.py
"""
SSE (Server-Sent Events) streaming helpers for GitHub Copilot agents.

GitHub Copilot expects responses in OpenAI streaming delta format:
    data: {"choices":[{"delta":{"content":"token"},"index":0}]}
    ...
    data: [DONE]
"""

import json
from typing import Generator


def sse_token(token: str) -> str:
    """
    Formats a single text token as an SSE data line.

    Args:
        token: The text content to send.

    Returns:
        A formatted SSE data line string.
    """
    chunk = json.dumps({"choices": [{"delta": {"content": token}, "index": 0}]})
    return f"data: {chunk}\n\n"


def sse_done() -> str:
    """Returns the final SSE [DONE] event."""
    return "data: [DONE]\n\n"


def sse_error(message: str) -> Generator[str, None, None]:
    """
    Yields an error message token followed by [DONE].

    Args:
        message: Human-readable error description.
    """
    yield sse_token(message)
    yield sse_done()


def pipe_openai_stream(response) -> Generator[str, None, None]:
    """
    Pipes an OpenAI-compatible streaming HTTP response as SSE lines.

    Works with any provider that returns `text/event-stream` in the same
    format as OpenAI (GitHub Models, Azure OpenAI, etc.).

    Args:
        response: A `requests.Response` object with `stream=True`.

    Yields:
        Raw SSE line strings to be written directly to the Flask response.
    """
    for line in response.iter_lines(decode_unicode=True):
        if line:
            yield line + "\n"
        else:
            yield "\n"

File: templates/nodejs-agent/src/stream.ts
import type { Response } from 'express';

/**
 * Writes the SSE headers required by GitHub Copilot.
 * Call this before sending any data chunks.
 */
export function startSSE(res: Response): void {
  res.setHeader('Content-Type', 'text/event-stream');
  res.setHeader('Cache-Control', 'no-cache');
  res.setHeader('Transfer-Encoding', 'chunked');
  res.setHeader('X-Accel-Buffering', 'no');
  res.flushHeaders();
}

/**
 * Sends a single content token as an OpenAI-compatible SSE delta.
 *
 * @param res   - Express response object
 * @param token - The text token to stream
 */
export function writeToken(res: Response, token: string): void {
  const chunk = JSON.stringify({
    choices: [{ delta: { content: token }, index: 0 }],
  });
  res.write(`data: ${chunk}\n\n`);
}

/**
 * Sends the final [DONE] event and ends the response.
 *
 * @param res - Express response object
 */
export function endSSE(res: Response): void {
  res.write('data: [DONE]\n\n');
  res.end();
}

/**
 * Pipes an OpenAI-compatible streaming response directly to the SSE client.
 * Works with any provider that returns `text/event-stream` (OpenAI, GitHub Models, Azure OpenAI).
 *
 * @param llmStream - ReadableStream from the LLM fetch response body
 * @param res       - Express response object (SSE headers must already be sent)
 */
export async function pipeStream(
  llmStream: ReadableStream<Uint8Array>,
  res: Response
): Promise<void> {
  const reader = llmStream.getReader();
  const decoder = new TextDecoder();

  try {
    while (true) {
      const { value, done } = await reader.read();
      if (done) break;

      const text = decoder.decode(value, { stream: true });
      // Forward each SSE line from the LLM directly to the client.
      for (const line of text.split('\n')) {
        if (line.trim()) {
          res.write(line + '\n');
        }
      }
      res.write('\n');
    }
  } finally {
    reader.releaseLock();
  }
}
